### PR TITLE
Switch ExperimentPage to use new filter field in SearchRuns

### PR DESCRIPTION
**Pull request by andrewmchen**
We added a new field in SearchRuns called filter which takes a string and parses the search filter on the server side. In this PR, we switch the UI component to use that field instead.

### CODE REVIEW

**Review by dbczumar — CHANGES_REQUESTED** [review]

**Comment by dbczumar** [comment]
Throughout the testing process, I found that the following queries produce UI errors and / or don't work, despite the fact that they work on master (regressions):

1. `“params.A = ’74’”`: This produces a niagara falls screen (note that the whole query is wrapped in quotation marks) due to the following exception with the code `INTERNAL_ERROR` (500):

```
{"message": "Invalid clause(s) in filter string: '\"params.A = '74'\"'", "error_code": "INTERNAL_ERROR"}
```

2. `params.A > 74`: This also produces niagara falls due to the following `INTERNAL_ERROR`:
```
{"message": "Invalid comparator '>' not one of 'set(['=', '!='])", "error_code": "INTERNAL_ERROR"}
```

3. `params.A = "74"`: This query correctly selects runs that have the (parameter, value) pair (`A`, `74`) on `master` (where the UI is using anded expressions). With the UI + filters API updates, this produces ```Invalid filter string '"74"'. Filter comparison is expected as 'metric.<key> <comparator> <value>' or'params.<key> <comparator> <value>'.```

Observations:

1. As I discussed with @andrewmchen, search APIs probably shouldn't return `INTERNAL_ERROR` in  the case of the above failures; `INVALID_PARAMETER_VALUE` is more appropriate and is explicitly handled by the UI. (Relevant stakeholder: @mparkhe)

2. The UI should more gracefully handle the `INTERNAL_ERROR` case. Producing a Niagara falls image is particularly bad because we save the search query as a browser cookie. As a result, refreshing the page attempts to invoke the same query and produces yet another Niagara falls view. The only way to escape "perpetual Niagara falls" is to clear the browser cache.

**Comment by mparkhe** [comment]
Some great catches @dbczumar. #1042 should address these.

File: mlflow/server/js/src/components/RunPage.js [hunk 1]
```diff
@@ -37,21 +37,23 @@ class RunPage extends Component {
...
-              return <RunNotFoundView runId={this.props.runUuid}/>;
-            }
-            return undefined;
-          }}
         >
-          <RunView
-            runUuid={this.props.runUuid}
-            getMetricPagePath={
-              (key) => Routes.getMetricPageRoute([this.props.runUuid], key, this.props.experimentId)
+          {(isLoading, shouldRenderError, requests) => {
```
Comment (andrewmchen) [inline mlflow/server/js/src/components/RunPage.js]: QAed this bit manually.

File: mlflow/server/js/src/components/ExperimentPage.js [hunk 2]
```diff
@@ -107,35 +106,48 @@ class ExperimentPage extends Component {
...
-              />);
+            if (shouldRenderError) {
+              const searchRunsRequest = Utils.getRequestWithId(
+                requests, this.state.searchRunsRequestId);
+              if (searchRunsRequest.error && searchRunsRequest.error.getErrorCode() ===
+                  ErrorCodes.INVALID_PARAMETER_VALUE) {
+                searchRunsError = searchRunsRequest.error.getMessageField();
+              } else if (getExperimentRequest.error.getErrorCode() ===
+                  ErrorCodes.PERMISSION_DENIED) {
+                return (<PermissionDeniedView
```
Comment (andrewmchen) [inline mlflow/server/js/src/components/ExperimentPage.js]: QAed this bit manually.

File: mlflow/server/js/src/Actions.js [hunk 3]
```diff
@@ -197,5 +197,20 @@ export class ErrorWrapper {
     }
     return "INTERNAL_SERVER_ERROR";
   }
+
+  getMessageField() {
+    const responseText = this.xhr.responseText;
+    if (responseText) {
+      try {
+        const parsed = JSON.parse(responseText);
+        if (parsed.error_code && parsed.message) {
```
[the remark this answers is not in the document]
Comment (andrewmchen) [reply mlflow/server/js/src/Actions.js]: It means that the response came from our OSS server in the way we expect it to (it should always have an `error_code` field and a `message` field.

File: mlflow/server/js/src/components/RequestStateWrapper.js [hunk 4]
```diff
@@ -1,20 +1,20 @@
...
 export class RequestStateWrapper extends Component {
   static propTypes = {
+    // Should this component render the child before all the requests are complete?
     shouldOptimisticallyRender: PropTypes.bool,
     requests: PropTypes.arrayOf(PropTypes.object).isRequired,
-    children: PropTypes.node.isRequired,
-    // (requests) => undefined | React Node.
-    // This function is called when all requests are complete and when one or more of them is
-    // in the error state. The function can choose to render an error view depending on the
+    // (shouldRenderError: boolean, requests) => undefined | React Node.
```
[the remark this answers is not in the document]
Comment (andrewmchen) [reply mlflow/server/js/src/components/RequestStateWrapper.js]: Correct!

**Comment by mparkhe** [comment]
@smurching and I tested with various different flavors. They all work with top of master

- `params."p name2" != ""`
- `params."p name2" = "a valid value"`
- `metrics.name > 0.23`
- `metrics."23_name_with_leading_number_quoted_key" > 0`

This is good to be merged in.

**Pull request merged**